test/cpp/tensorexpr/test_loopnest.cpp

```diff
@@ -546,5 +547,224 @@ void testScheduleDynamicShape2D() {
   testWithSize(37, 11);
 }
 
+static std::unordered_map<const Buf*, TensorAccessBoundsInfo>
+convertBoundsInfoToMap(const std::vector<TensorAccessBoundsInfo>& v) {
+  std::unordered_map<const Buf*, TensorAccessBoundsInfo> res;
+  for (const auto& el : v) {
+    res[el.buf] = el;
+  }
+  return res;
+}
+
+static void verifyConstBounds(
+    const TensorAccessBoundsInfo& access_info,
+    const std::vector<std::pair<int, int>>& ref) {
+  size_t ndim = ref.size();
+  ASSERT_EQ(access_info.start.size(), ndim);
+  ASSERT_EQ(access_info.stop.size(), ndim);
+  for (size_t i = 0; i < ndim; i++) {
+    if (ref[i].first >= 0) { // Negative values are used to skip the check
+      auto start_imm = dynamic_cast<const IntImm*>(access_info.start[i]);
+      ASSERT_TRUE(start_imm);
+      ASSERT_EQ(start_imm->value(), ref[i].first);
+    }
+    if (ref[i].second >= 0) {
+      auto stop_imm = dynamic_cast<const IntImm*>(access_info.stop[i]);
+      ASSERT_TRUE(stop_imm);
+      ASSERT_EQ(stop_imm->value(), ref[i].second);
+    }
+  }
+}
+
+void testBoundsInference_1() {
+  // Verify that bounds inference works for the following example:
+  // for i in 0..100:
+  //   b[i] = a[i]
+  // For this loop bounds inference should yield the following:
+  // {{b, kStore, 0, 99}, {a, kLoad, 0, 99}}
```

Comment:
Minor: as for the bound, I mild prefer the C++ convention: left-close, right-open. So the same range should be: [0, 100)

Reply:
It is actually quite inconvenient in this case to use that convention the reason for that is how bounds inference works: it starts from a scope of Load/Store/FunctionCall and records indices used there and then escapes to outer scopes one by one replacing start/stop values for the corresponding index with Min/Max values of the iteration range. E.g:
```
(1)  for x in 0..10:
(2)    a[x] = b[x*2]
```
In this loopnest will start from (2) and infer bounds: `b: [x*2; x*2], a: [x; x]`. Then we will go to the scope of the loop (1). We will take the inferred bounds and in them we will replace `x` with `0` for the start value and with `10-1` (the last taken value, which equals stop values minus 1) for the stop value: `b: [0;18], a:[0;9]`. We could add +1 everywhere to follow the `[start;stop)` notation, but we would still have to do this `-1` and then `+1` at every step: i.e. we cannot just replace `x` with `10` (stop value), as it would give us incorrect bounds `b:[0:20), a:[0:10)` (correct bounds for `b` would be `[0:19)`).